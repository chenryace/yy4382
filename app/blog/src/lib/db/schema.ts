import { sql } from "drizzle-orm";
import { serial, timestamp } from "drizzle-orm/pg-core";
import {
  check,
  integer,
  pgTable,
  text,
  boolean,
  type AnyPgColumn,
} from "drizzle-orm/pg-core";

export const user = pgTable("user", {
  id: text("id").primaryKey(),
  name: text("name").notNull(),
  email: text("email").notNull().unique(),
  emailVerified: boolean("email_verified").notNull().default(false),
  image: text("image"),
  createdAt: timestamp("created_at").notNull().defaultNow(),
  updatedAt: timestamp("updated_at").notNull().defaultNow(),
  // `admin` plugin fields
  role: text("role"),
  banned: boolean("banned"),
  banReason: text("ban_reason"),
  banExpires: timestamp("ban_expires"),
});

export const session = pgTable("session", {
  id: text("id").primaryKey(),
  expiresAt: timestamp("expires_at").notNull(),
  token: text("token").notNull().unique(),
  createdAt: timestamp("created_at").notNull(),
  updatedAt: timestamp("updated_at").notNull(),
  ipAddress: text("ip_address"),
  userAgent: text("user_agent"),
  userId: text("user_id")
    .notNull()
    .references(() => user.id, { onDelete: "cascade" }),

  // `admin` plugin fields
  impersonatedBy: text("impersonated_by"),
});

export const account = pgTable("account", {
  id: text("id").primaryKey(),
  accountId: text("account_id").notNull(),
  providerId: text("provider_id").notNull(),
  userId: text("user_id")
    .notNull()
    .references(() => user.id, { onDelete: "cascade" }),
  accessToken: text("access_token"),
  refreshToken: text("refresh_token"),
  idToken: text("id_token"),
  accessTokenExpiresAt: timestamp("access_token_expires_at"),
  refreshTokenExpiresAt: timestamp("refresh_token_expires_at"),
  scope: text("scope"),
  password: text("password"),
  createdAt: timestamp("created_at").notNull(),
  updatedAt: timestamp("updated_at").notNull(),
});

export const verification = pgTable("verification", {
  id: text("id").primaryKey(),
  identifier: text("identifier").notNull(),
  value: text("value").notNull(),
  expiresAt: timestamp("expires_at").notNull(),
  createdAt: timestamp("created_at").notNull().defaultNow(),
  updatedAt: timestamp("updated_at").notNull().defaultNow(),
});

export const comment = pgTable(
  "comment",
  {
    id: serial("id").primaryKey(),

    rawContent: text("raw_content").notNull(),
    renderedContent: text("rendered_content").notNull(),

    path: text("path").notNull(),
    parentId: integer("parent_id").references((): AnyPgColumn => comment.id),
    replyToId: integer("reply_to_id").references((): AnyPgColumn => comment.id),

    createdAt: timestamp("created_at").notNull().defaultNow(),
    updatedAt: timestamp("updated_at").notNull().defaultNow(),
    deletedAt: timestamp("deleted_at"),

    userId: text("user_id").references(() => user.id),
    userIp: text("user_ip"),
    userAgent: text("user_agent"),

    visitorName: text("visitor_name"),
    visitorEmail: text("visitor_email"),

    anonymousName: text("anonymous_name"),

    isSpam: boolean("is_spam").notNull(),
  },
  (table) => [
    check(
      "either_id_or_visitor_name",
      sql`(${table.userId} IS NULL AND ${table.visitorName} IS NOT NULL) OR (${table.userId} IS NOT NULL AND ${table.visitorName} IS NULL)`,
    ),
    check(
      "no_visitor_email_if_no_name",
      sql`${table.visitorName} IS NOT NULL OR (${table.visitorEmail} IS NULL)`,
    ),
  ],
);
