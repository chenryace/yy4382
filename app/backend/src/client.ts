import { hc } from "hono/client";
import type { App } from ".";

export function createClient(...params: Parameters<typeof hc>) {
  return hc<App>(...params).api.v1;
}
export type { App as BackendAppType };
export * from "hono/client";
