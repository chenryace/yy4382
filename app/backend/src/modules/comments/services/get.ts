import type { User } from "@/auth/auth-plugin.js";
import type { DbClient } from "@/db/db-plugin.js";
import type {
  GetCommentsBody,
  LayeredCommentList,
  LayeredCommentData,
} from "@repo/api/comment/get.model";
import { and, asc, desc, eq, inArray, isNull, or, sql } from "drizzle-orm";
import { comment, user } from "@/db/schema.js";
import { tablesToCommentData } from "./comment-data.js";
import type { CommentData } from "@repo/api/comment/comment-data";

export async function getComments(
  body: GetCommentsBody,
  options: { db: DbClient; user: User | null },
): Promise<{ comments: LayeredCommentList; total: number }> {
  const { path, limit, offset, sortBy } = body;
  const { db, user } = options;
  const { comments } = await getCommentsDb(db, user, {
    path,
    limit,
    offset,
    sortBy,
  });
  const layeredComments = layerComments(comments, sortBy);
  const totalCount = await db.$count(
    comment,
    and(
      isNull(comment.deletedAt),
      eq(comment.isSpam, false),
      path === null ? sql`1=1` : eq(comment.path, path),
      isNull(comment.parentId),
    ),
  );

  return { comments: layeredComments, total: totalCount };
}

async function getCommentsDb(
  db: DbClient,
  currentUser: User | null,
  {
    path,
    limit,
    offset,
    sortBy,
  }: {
    path: string | null;
    limit: number;
    offset: number;
    sortBy: GetCommentsBody["sortBy"];
  },
): Promise<{ comments: CommentData[] }> {
  const rootCommentsWith = db.$with("root_comments").as(
    db
      .select({ id: comment.id })
      .from(comment)
      .where(
        and(
          path === null ? sql`1=1` : eq(comment.path, path),
          isNull(comment.parentId),
          isNull(comment.deletedAt),
          eq(comment.isSpam, false),
        ),
      )
      .orderBy(
        sortBy === "created_desc"
          ? desc(comment.createdAt)
          : asc(comment.createdAt),
      )
      .limit(limit)
      .offset(offset),
  );
  const comments = await db
    .with(rootCommentsWith)
    .select()
    .from(comment)
    .leftJoin(user, eq(comment.userId, user.id))
    .where(
      or(
        inArray(comment.id, sql`(SELECT id FROM ${rootCommentsWith})`),
        and(
          inArray(comment.parentId, sql`(SELECT id FROM ${rootCommentsWith})`),
          isNull(comment.deletedAt),
          eq(comment.isSpam, false),
          path === null ? sql`1=1` : eq(comment.path, path),
        ),
      ),
    );
  return {
    comments: comments.map((comment) =>
      tablesToCommentData(
        comment.comment,
        comment.user,
        currentUser?.role === "admin",
      ),
    ),
  };
}

function layerComments(
  comments: CommentData[],
  sortBy: GetCommentsBody["sortBy"],
): LayeredCommentData[] {
  const map = new Map<
    number,
    LayeredCommentData | { children: CommentData[] }
  >();

  for (const comment of comments) {
    if (comment.parentId === null) {
      map.set(comment.id, {
        ...comment,
        children: map.get(comment.id)?.children ?? [],
      });
    } else {
      const parent = map.get(comment.parentId);
      if (parent) {
        parent.children.push(comment);
      } else {
        map.set(comment.parentId, {
          children: [],
        });
      }
    }
  }

  return Array.from(map.values())
    .filter((value): value is LayeredCommentData => "id" in value)
    .sort((a, b) =>
      sortBy === "created_desc"
        ? new Date(b.createdAt).getTime() - new Date(a.createdAt).getTime()
        : new Date(a.createdAt).getTime() - new Date(b.createdAt).getTime(),
    )
    .map((entry) => {
      return {
        ...entry,
        children: entry.children.sort(
          (a, b) =>
            new Date(a.createdAt).getTime() - new Date(b.createdAt).getTime(),
        ),
      };
    });
}
