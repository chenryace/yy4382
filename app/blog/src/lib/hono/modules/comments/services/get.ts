import type { User } from "@/lib/auth/auth-plugin";
import type { DbClient } from "@/lib/db/db-plugin";
import type {
  GetCommentsBody,
  CommentData,
  LayeredCommentList,
  LayeredCommentData,
} from "./get.model";
import { and, asc, desc, eq, inArray, isNull, or, sql } from "drizzle-orm";
import { comment, user } from "@/lib/db/schema";
import crypto from "node:crypto";

export async function getComments(
  body: GetCommentsBody,
  options: { db: DbClient; user: User | null },
): Promise<{ comments: LayeredCommentList; total: number }> {
  const { path, limit, offset, sortBy } = body;
  const { db, user } = options;
  const { comments } = await getCommentsDb(db, user, {
    path,
    limit,
    offset,
    sortBy,
  });
  const layeredComments = layerComments(comments, sortBy);
  const totalCount = await db.$count(
    comment,
    and(
      isNull(comment.deletedAt),
      eq(comment.isSpam, false),
      path === null ? sql`1=1` : eq(comment.path, path),
      isNull(comment.parentId),
    ),
  );

  return { comments: layeredComments, total: totalCount };
}

async function getCommentsDb(
  db: DbClient,
  currentUser: User | null,
  {
    path,
    limit,
    offset,
    sortBy,
  }: {
    path: string | null;
    limit: number;
    offset: number;
    sortBy: GetCommentsBody["sortBy"];
  },
): Promise<{ comments: CommentData[] }> {
  const rootCommentsWith = db.$with("root_comments").as(
    db
      .select({ id: comment.id })
      .from(comment)
      .where(
        and(
          path === null ? sql`1=1` : eq(comment.path, path),
          isNull(comment.parentId),
          isNull(comment.deletedAt),
          eq(comment.isSpam, false),
        ),
      )
      .orderBy(
        sortBy === "created_desc"
          ? desc(comment.createdAt)
          : asc(comment.createdAt),
      )
      .limit(limit)
      .offset(offset),
  );
  const comments = await db
    .with(rootCommentsWith)
    .select({
      id: comment.id,
      content: comment.renderedContent,
      rawContent: comment.rawContent,
      path: comment.path,

      parentId: comment.parentId,
      replyToId: comment.replyToId,

      createdAt: comment.createdAt,
      updatedAt: comment.updatedAt,

      userId: comment.userId,
      userIp: comment.userIp,
      userAgent: comment.userAgent,
      userName: user.name,
      userEmail: user.email,
      userImage: user.image,

      visitorName: comment.visitorName,
      visitorEmail: comment.visitorEmail,
      anonymousName: comment.anonymousName,
    })
    .from(comment)
    .leftJoin(user, eq(comment.userId, user.id))
    .where(
      or(
        inArray(comment.id, sql`(SELECT id FROM ${rootCommentsWith})`),
        and(
          inArray(comment.parentId, sql`(SELECT id FROM ${rootCommentsWith})`),
          isNull(comment.deletedAt),
          eq(comment.isSpam, false),
          path === null ? sql`1=1` : eq(comment.path, path),
        ),
      ),
    );
  const adminCommentData: CommentData[] = comments.map((comment) => {
    const sanitized = {
      ...comment,
      createdAt: comment.createdAt,
      updatedAt: comment.updatedAt,
      displayName:
        (comment.anonymousName || comment.visitorName || comment.userName) ??
        "Unknown",
      userImage: comment.anonymousName
        ? "https://avatar.vercel.sh/anonymous"
        : (comment.userImage ??
          getGravatarUrl(comment.userEmail ?? comment.visitorEmail ?? "")),
    };
    return sanitized;
  });

  const sanitizedComments: CommentData[] =
    currentUser?.role === "admin"
      ? adminCommentData
      : adminCommentData.map((c): CommentData => {
          const data = {
            id: c.id,
            content: c.content,
            rawContent: c.rawContent,
            parentId: c.parentId,
            replyToId: c.replyToId,
            createdAt: c.createdAt,
            updatedAt: c.updatedAt,
            userImage: c.userImage,
            displayName: c.displayName,
            path: c.path,
            userId: c.anonymousName ? null : c.userId,
          };
          return data;
        });
  return { comments: sanitizedComments };
}

function getGravatarUrl(
  email: string,
  size = 200,
  defaultImage = "identicon",
  rating = "g",
) {
  const cleanEmail = email.trim().toLowerCase();
  const hash = crypto.createHash("md5").update(cleanEmail).digest("hex");
  return `https://www.gravatar.com/avatar/${hash}?s=${size}&d=${defaultImage}&r=${rating}`;
}

function layerComments(
  comments: CommentData[],
  sortBy: GetCommentsBody["sortBy"],
): LayeredCommentData[] {
  const map = new Map<
    number,
    LayeredCommentData | { children: CommentData[] }
  >();

  for (const comment of comments) {
    if (comment.parentId === null) {
      map.set(comment.id, {
        ...comment,
        children: map.get(comment.id)?.children ?? [],
      });
    } else {
      const parent = map.get(comment.parentId);
      if (parent) {
        parent.children.push(comment);
      } else {
        map.set(comment.parentId, {
          children: [],
        });
      }
    }
  }

  return Array.from(map.values())
    .filter((value): value is LayeredCommentData => "id" in value)
    .sort((a, b) =>
      sortBy === "created_desc"
        ? new Date(b.createdAt).getTime() - new Date(a.createdAt).getTime()
        : new Date(a.createdAt).getTime() - new Date(b.createdAt).getTime(),
    )
    .map((entry) => {
      return {
        ...entry,
        children: entry.children.sort(
          (a, b) =>
            new Date(a.createdAt).getTime() - new Date(b.createdAt).getTime(),
        ),
      };
    });
}
