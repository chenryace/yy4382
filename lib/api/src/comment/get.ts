import { type Result, err, ok } from "@repo/helpers/result";
import {
  type GetCommentsBody,
  type GetCommentsChildrenBody,
  getCommentsChildrenResponse,
  type GetCommentsChildrenResponse,
  type GetCommentsResponse,
  getCommentsResponse,
} from "./get.model.js";

export async function getComments(
  options: GetCommentsBody,
  serverUrl: string,
): Promise<Result<GetCommentsResponse, string>> {
  const path = new URL("v1/comments/get", serverUrl).href;
  const resp = await fetch(path, {
    method: "POST",
    body: JSON.stringify(options),
    headers: {
      "Content-Type": "application/json",
    },
    credentials: "include",
  });
  if (!resp.ok) {
    return err(await resp.text());
  }
  return ok(getCommentsResponse.parse(await resp.json()));
}

export async function getCommentsChildren(
  options: GetCommentsChildrenBody,
  serverUrl: string,
): Promise<Result<GetCommentsChildrenResponse, string>> {
  const path = new URL("v1/comments/get-children", serverUrl).href;
  const resp = await fetch(path, {
    method: "POST",
    body: JSON.stringify(options),
    headers: {
      "Content-Type": "application/json",
    },
    credentials: "include",
  });
  if (!resp.ok) {
    return err(await resp.text());
  }
  return ok(getCommentsChildrenResponse.parse(await resp.json()));
}
