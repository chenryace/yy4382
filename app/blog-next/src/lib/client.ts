import { env } from "@/env";
import { createAuthClient } from "better-auth/react";
import { adminClient, magicLinkClient } from "better-auth/client/plugins";

export const authClient = createAuthClient({
  baseURL: new URL("v1/auth", env.NEXT_PUBLIC_BACKEND_URL).href,
  plugins: [adminClient(), magicLinkClient()],
});

export type Session = typeof authClient.$Infer.Session;
