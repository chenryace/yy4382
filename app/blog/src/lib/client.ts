import type { App } from "@/app/api/[[...route]]/route";
import { hc } from "hono/client";
import { createAuthClient } from "better-auth/react";
import { adminClient } from "better-auth/client/plugins";

export const honoClient = hc<App>("/", {
  fetch: (input: RequestInfo | URL, init?: RequestInit) => {
    return fetch(input, {
      ...init,
      credentials: "include",
    });
  },
}).api.v1;

export const authClient = createAuthClient({
  basePath: "/api/v1/auth",
  plugins: [adminClient()],
});

export type Session = typeof authClient.$Infer.Session;
