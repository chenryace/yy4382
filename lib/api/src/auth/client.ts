import { createAuthClient } from "better-auth/react";
import { adminClient, magicLinkClient } from "better-auth/client/plugins";

export const authClient = (serverUrl: string) =>
  createAuthClient({
    baseURL: new URL("v1/auth", serverUrl).href,
    plugins: [adminClient(), magicLinkClient()],
  });

export type Session = Awaited<
  ReturnType<typeof authClient>
>["$Infer"]["Session"]["session"];
export type User = Awaited<
  ReturnType<typeof authClient>
>["$Infer"]["Session"]["user"];
