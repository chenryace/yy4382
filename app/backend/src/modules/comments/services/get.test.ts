/**
 * This test file aims to test the `getComments` func in get-comment service.
 *
 * Should be tested:
 * - returns proper rows from db
 *   - take into account: spam, deleted, path, user role, etc.
 * - layered comments are properly created
 * - pagination works properly
 * - sorting works properly
 *
 * Should not be tested:
 * - which fields in a row should be returned to user
 *   - due to user role difference, some fields are only returned to admin or to owner
 *   - this is achieved by a shared function `tablesToCommentData`
 *   - should test it separately
 */

import { describe, it, expect, vi, beforeEach, afterEach } from "vitest";
import { drizzle } from "drizzle-orm/libsql";
import { getComments } from "./get.js";
import * as schema from "@/db/schema.js";
import { eq } from "drizzle-orm";
import type { DbClient } from "@/db/db-plugin.js";
import { migrate } from "drizzle-orm/libsql/migrator";
import { createClient } from "@libsql/client";
import type { Client } from "@libsql/client";

let client: Client;
let db: DbClient;

const { user, comment } = schema;

const mockUser: (typeof user.$inferInsert)[] = [
  {
    id: "1",
    name: "admin",
    email: "admin@example.com",
    role: "admin",
  },
  {
    id: "2",
    name: "user1",
    email: "user1@example.com",
    role: "user",
  },
  {
    id: "3",
    name: "user2",
    email: "user2@example.com",
    role: "user",
  },
];

const makeComment = (
  args: Partial<typeof comment.$inferInsert>,
): typeof comment.$inferInsert => {
  if (!args.rawContent) {
    throw new Error("rawContent is required");
  }
  return {
    path: "/",
    isSpam: false,
    rawContent: args.rawContent,
    renderedContent: args.rawContent,
    ...args,
  };
};

const makeMockComment = () => {
  const mockComment: (typeof comment.$inferInsert)[] = [];
  let count = 1;
  for (const isSpam of [false, true]) {
    for (const anonymous of [false, true]) {
      for (const userData of mockUser) {
        const comment = makeComment({
          id: count++,
          rawContent: `comment ${count} from "${userData.name}" role "${userData.role}" isSpam "${isSpam}" anonymous "${anonymous}"`,
          createdAt: new Date(count * 10000),
          isSpam,
          userId: userData.id,
        });
        if (anonymous) {
          comment.anonymousName = "Anonymous";
        }
        mockComment.push(comment);
      }
      mockComment.push(
        makeComment({
          id: count++,
          rawContent: `comment ${count} from "Visitor" role "visitor" isSpam "${isSpam}" anonymous "${anonymous}"`,
          visitorName: "Visitor",
          visitorEmail: "visitor@example.com",
          createdAt: new Date(count * 10000),
          isSpam,
          anonymousName: anonymous ? "Anonymous" : undefined,
        }),
      );
    }
  }

  const parentId = count;
  const parentComment = makeComment({
    id: parentId,
    rawContent: `comment ${parentId} will be parent`,
    createdAt: new Date(parentId * 10000),
    userId: "1",
  });
  mockComment.push(parentComment);
  count++;
  const firstReplyId = count;
  for (let i = 0; i < 2; i++) {
    const comment = makeComment({
      id: count++,
      rawContent: `reply ${count} to comment ${parentId}`,
      createdAt: new Date(count * 10000),
      parentId,
      replyToId: parentId,
      userId: "1",
    });
    mockComment.push(comment);
  }
  mockComment.push(
    makeComment({
      id: count++,
      rawContent: `reply to reply to comment ${parentId}`,
      createdAt: new Date(count * 10000),
      parentId,
      replyToId: firstReplyId,
      userId: "1",
    }),
  );
  mockComment.push(
    makeComment({
      id: count++,
      rawContent: `reply to reply to reply to comment ${parentId}`,
      createdAt: new Date(count * 10000),
      parentId,
      replyToId: firstReplyId,
      deletedAt: new Date(count * 1000000),
      userId: "1",
    }),
  );
  return mockComment;
};

const mockComment: (typeof comment.$inferInsert)[] = makeMockComment();

beforeEach(async () => {
  client = createClient({ url: ":memory:" });
  db = drizzle({ client, schema });
  await migrate(db, { migrationsFolder: "./drizzle" });
  vi.setSystemTime(new Date("2000-01-01T00:00:00.000Z"));

  await db.insert(user).values(mockUser);
  await db.insert(comment).values(mockComment);
});

afterEach(() => {
  vi.useRealTimers();
});

describe("admin get comments", () => {
  it("should return comments", async () => {
    const admin = (await db.select().from(user).where(eq(user.id, "1")))[0];
    const { comments, total } = await getComments(
      { path: "/", limit: 1000, offset: 0, sortBy: "created_asc" },
      { db, user: admin },
    );
    expect(total).toBe(17);

    expect(
      comments.find((c) => c.children.length !== 0)?.children?.length,
      "parent comment has proper amount of children",
    ).toBe(3);
    expect(
      comments.at(0),
      "admin comment should include all info",
    ).toMatchSnapshot();
  });
  it("should include spam comments", async () => {
    const admin = (await db.select().from(user).where(eq(user.id, "1")))[0];
    const { comments } = await getComments(
      { path: "/", limit: 1000, offset: 0, sortBy: "created_asc" },
      { db, user: admin },
    );
    expect(comments.find((c) => c.isSpam)).toBeDefined();
  });

  it("should return comments with limit and offset", async () => {
    const admin = (await db.select().from(user).where(eq(user.id, "1")))[0];
    const { comments } = await getComments(
      { path: "/", limit: 5, offset: 5, sortBy: "created_asc" },
      { db, user: admin },
    );
    expect(comments.length).toBe(5);
    expect(comments[0].id).toBe(6);
  });
  it("should return comments with sortBy", async () => {
    const admin = (await db.select().from(user).where(eq(user.id, "1")))[0];
    const { comments } = await getComments(
      { path: "/", limit: 2, offset: 2, sortBy: "created_desc" },
      { db, user: admin },
    );
    expect(comments.length).toBe(2);
    expect(comments[0].id).toBeGreaterThan(comments[1].id);
  });
});

describe("non-admin get comments", () => {
  it("should return comments", async () => {
    const user1 = (await db.select().from(user).where(eq(user.id, "2")))[0];
    const { comments, total } = await getComments(
      { path: "/", limit: 1000, offset: 0, sortBy: "created_asc" },
      { db, user: user1 },
    );
    expect(total).toBe(9);

    expect(
      comments.find((c) => c.children.length !== 0)?.children?.length,
      "parent comment has proper amount of children",
    ).toBe(3);
    expect(
      comments.at(0),
      "non-admin comment should include partial info",
    ).toMatchSnapshot();
  });
  it("should not include spam comments", async () => {
    const user1 = (await db.select().from(user).where(eq(user.id, "2")))[0];
    const { comments } = await getComments(
      { path: "/", limit: 1000, offset: 0, sortBy: "created_asc" },
      { db, user: user1 },
    );
    expect(comments.find((c) => c.isSpam)).toBeUndefined();
  });
});
