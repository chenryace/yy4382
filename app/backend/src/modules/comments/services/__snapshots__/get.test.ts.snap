// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`cursor style pagination > should return first page correctly 1`] = `
{
  "comments": [
    {
      "children": {
        "cursor": 41,
        "data": [
          {
            "id": 41,
          },
        ],
        "hasMore": false,
        "total": 1,
      },
      "data": {
        "id": 4,
      },
    },
    {
      "children": {
        "cursor": 0,
        "data": [],
        "hasMore": false,
        "total": 0,
      },
      "data": {
        "id": 3,
      },
    },
  ],
  "cursor": 3,
  "hasMore": true,
  "total": 12,
}
`;

exports[`cursor style pagination > should return second page correctly 1`] = `
{
  "comments": [
    {
      "children": {
        "cursor": 23,
        "data": [
          {
            "id": 21,
          },
          {
            "id": 22,
          },
          {
            "id": 23,
          },
        ],
        "hasMore": false,
        "total": 3,
      },
      "data": {
        "id": 2,
      },
    },
    {
      "children": {
        "cursor": 13,
        "data": [
          {
            "id": 11,
          },
          {
            "id": 12,
          },
          {
            "id": 13,
          },
        ],
        "hasMore": true,
        "total": 4,
      },
      "data": {
        "id": 1,
      },
    },
  ],
  "cursor": 1,
  "hasMore": false,
  "total": 12,
}
`;

exports[`cursor style pagination > should work with created_asc 1`] = `
{
  "comments": [
    {
      "children": {
        "cursor": 13,
        "data": [
          {
            "id": 11,
          },
          {
            "id": 12,
          },
          {
            "id": 13,
          },
        ],
        "hasMore": true,
        "total": 4,
      },
      "data": {
        "id": 1,
      },
    },
    {
      "children": {
        "cursor": 23,
        "data": [
          {
            "id": 21,
          },
          {
            "id": 22,
          },
          {
            "id": 23,
          },
        ],
        "hasMore": false,
        "total": 3,
      },
      "data": {
        "id": 2,
      },
    },
    {
      "children": {
        "cursor": 0,
        "data": [],
        "hasMore": false,
        "total": 0,
      },
      "data": {
        "id": 3,
      },
    },
  ],
  "cursor": 3,
  "hasMore": true,
  "total": 12,
}
`;

exports[`get more children with cursor > should return first page correctly 1`] = `
{
  "cursor": 16,
  "data": [
    {
      "id": 17,
    },
    {
      "id": 16,
    },
  ],
  "hasMore": true,
}
`;

exports[`get more children with cursor > should return second page correctly 1`] = `
{
  "cursor": 14,
  "data": [
    {
      "id": 15,
    },
    {
      "id": 14,
    },
  ],
  "hasMore": true,
}
`;

exports[`get more children with cursor > should work with created_asc 1`] = `
{
  "cursor": 13,
  "data": [
    {
      "id": 11,
    },
    {
      "id": 12,
    },
    {
      "id": 13,
    },
  ],
  "hasMore": true,
}
`;
